### PR TITLE
Support preview GraphQL API v4 Node IDs

### LINKED ISSUE #814: Support preview GraphQL API v4 Node IDs in REST API v3 resources
GitHub Developer API announcement:
https://developer.github.com/changes/2017-12-19-graphql-node-id/

To complete this issue:

* a new media type (`application/vnd.github.jean-grey-preview+json`) must be added to `github.go`,
* this media type must be added to the header of calls to all affected endpoints (see announcement),
* a new `node_id` field must be added to the appropriate responses structs that are returned.

This issue presents a great learning opportunity for a new contributor to participate and interact with this community of passionate and enthusiastic Go developers. Any and all contributions are greatly appreciated!

If you would like to work on this issue, please comment below and our team will send you an invite so we can assign the issue to you. Make sure to check out the [CONTRIBUTING.md](https://github.com/google/go-github/blob/master/CONTRIBUTING.md) documentation for help on getting started. Thank you!

### PR DESCRIPTION
**Pull request by anubhakushwaha**
Helps #814 .
The current pull request includes the GraphQL node ID in the response for the following REST API v3 resources:

- Deployments
- Gists
- Git Blobs
- Git Commits
- Git References
- Git Tags
- Issues
- Labels
- Milestones
- Organizations

### CODE REVIEW

**Comment by anubhakushwaha** [comment]
I signed it!

**Comment by anubhakushwaha** [comment]
@gmlewis  @shurcooL  I have added GraphQL node ID in the response, let me know if I am doing it correctly and would love to correct and work upon my mistakes.Thanks!

File: github/gists.go [hunk 1]
```diff
@@ -30,6 +30,7 @@ type Gist struct {
 	GitPushURL  *string                   `json:"git_push_url,omitempty"`
 	CreatedAt   *time.Time                `json:"created_at,omitempty"`
 	UpdatedAt   *time.Time                `json:"updated_at,omitempty"`
+	NodeID	    *string		  `json:"node_id,omitempty"`
```
Comment (shurcooL) [inline github/gists.go]: This line and many others aren't gofmted. You'll want to use [`gofmt`](https://golang.org/cmd/gofmt/) command to format all these files. You can either invoke it directly in this directory:

```
gofmt -l -w .
```

Or you can format the entire package via `go`'s `fmt` subcommand:

https://golang.org/cmd/go/#hdr-Run_gofmt_on_package_sources

```
go fmt github.com/google/go-github/github
```

File: github/git_blobs_test.go [hunk 2]
```diff
@@ -70,6 +75,10 @@ func TestGitService_CreateBlob(t *testing.T) {
 			t.Errorf("Request method = %v, want %v", r.Method, m)
 		}
 
+		if m := mediaTypeGraphQLNodeIDPreview; m != r.Header.Get("Accept") {
+			t.Errorf("Request accept header = %v, want %v", r.Header.Get("Accept"), m)
+		}
```
Comment (shurcooL) [inline github/git_blobs_test.go]: Here as well.

File: github/git_commits_test.go [hunk 3]
```diff
@@ -12,15 +12,17 @@ import (
...
 
 func TestGitService_GetCommit(t *testing.T) {
 	client, mux, _, teardown := setup()
 	defer teardown()
 
+	acceptHeaders := []string{mediaTypeGitSigningPreview, mediaTypeGraphQLNodeIDPreview}
 	mux.HandleFunc("/repos/o/r/git/commits/s", func(w http.ResponseWriter, r *http.Request) {
 		testMethod(t, r, "GET")
-		testHeader(t, r, "Accept", mediaTypeGitSigningPreview)
+		testHeader(t, r, "Accept", strings.Join(acceptHeaders, ", "))
```
Comment (shurcooL) [inline github/git_commits_test.go]: This is minor and either way is fine, but I'll make the suggestion for you to consider.

Since `acceptHeaders` is only used here, you can declare it right here. No need to declare it that much earlier:

```Go
acceptHeaders := []string{mediaTypeGitSigningPreview, mediaTypeGraphQLNodeIDPreview}
testHeader(t, r, "Accept", strings.Join(acceptHeaders, ", "))
```

**Comment by shurcooL** [comment]
Also, you should edit the PR description and replace "Partially fixes #814." with something like "Helps #814." GitHub [syntax](https://github.com/blog/1506-closing-issues-via-pull-requests) doesn't recognize the word "partially", it only sees "fixes #814", so it'll close that issue when this is merged, which isn't what you want.

File: github/gists.go (hunk 1 above)
Comment (anubhakushwaha) [reply github/gists.go]: Thanks @shurcooL  Resolved!

File: github/gists.go [hunk 4]
```diff
@@ -193,6 +209,9 @@ func (s *GistsService) GetRevision(ctx context.Context, id, sha string) (*Gist,
 	if err != nil {
 		return nil, nil, err
 	}
```
[the remark this answers is not in the document]
Comment (anubhakushwaha) [reply github/gists.go]: Done

**Comment by kshitij10496** [comment]
The build is failing due to formatting issues with the code.

Run `$ gofmt -d -s .`in a terminal from the base directory of the repo.
This will show the diff of all formatting issues with the code.
I think fixing them should resolve the build failure.

File: github/git_blobs_test.go [hunk 5]
```diff
@@ -22,6 +22,11 @@ func TestGitService_GetBlob(t *testing.T) {
 		if m := "GET"; m != r.Method {
 			t.Errorf("Request method = %v, want %v", r.Method, m)
 		}
+
+		if m := mediaTypeGraphQLNodeIDPreview; m != r.Header.Get("Accept") {
+			t.Errorf("Request accept header = %v, want %v", r.Header.Get("Accept"), m)
+		}
```
[the remark this answers is not in the document]
Comment (anubhakushwaha) [reply github/git_blobs_test.go]: @shurcooL  While working on this file, I noticed that in [git_blobs_test.go](https://github.com/google/go-github/blob/master/github/git_blobs_test.go#L22) and [git_trees_test.go](https://github.com/google/go-github/blob/master/github/git_trees_test.go#L22) we have used:
> 		if m := "GET"; m != r.Method {
>			t.Errorf("Request method = %v, want %v", r.Method, m)
>              }
which is tested differently from the rest of the files for the header where we have done it like :
>		testMethod(t, r, "GET")
 Is their any specific reason for doing so? I implemented the test on the **accept header** to maintain the consistency with the respective files. Let me know your thoughts.

File: github/git_blobs_test.go (hunk 2 above)
Comment (anubhakushwaha) [reply github/git_blobs_test.go]: Queried above.

File: github/git_commits_test.go (hunk 3 above)
Comment (anubhakushwaha) [reply github/git_commits_test.go]: @shurcooL I added it this way so that in future if there is a need for more **mediaType** values we can just add it to the list, what do you think? Else as per your say I will change it :smile:

**Comment by anubhakushwaha** [comment]
@kshitij10496  Thanks, I will keep it in mind from future :smile:

File: github/issues_labels.go [hunk 6]
```diff
@@ -82,6 +82,9 @@ func (s *IssuesService) CreateLabel(ctx context.Context, owner string, repo stri
 		return nil, nil, err
 	}
 
+	// TODO: remove custom Accept header when this API fully launches.
+	req.Header.Set("Accept", mediaTypeGraphQLNodeIDPreview)
```
Comment (anubhakushwaha) [inline github/issues_labels.go]: One of the reasons of failing Travis, forgot to add it previously.

File: github/git_commits_test.go (hunk 3 above)
Comment (anubhakushwaha) [reply github/git_commits_test.go]: @shurcooL  It was done to maintain consistency with the rest of the files when we had multiple mediaType values. Let me know if I should change it :smile:

File: github/git_blobs_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sahildua2305) [reply github/git_blobs_test.go]: @anubhakushwaha IMO, the inconsistency in `git_trees_test.go` and `git_blobs_test.go` is by accident rather than having any reasoning behind it. I'll suggest you to use `testHeader` in your PR to make it consistent with the entire package.

We should instead fix the inconsistency to avoid confusions like this one in future.

[the remark this answers is not in the document]
Comment (anubhakushwaha) [reply github/git_blobs_test.go]: @sahildua2305  Thanks for the suggestion, will update the PR.

File: github/git_commits_test.go (hunk 3 above)
Comment (shurcooL) [reply github/git_commits_test.go]: Okay, keeping it as is is completely acceptable.

File: github/git_commits.go [hunk 7]
```diff
@@ -35,7 +36,8 @@ type Commit struct {
 	// CommentCount is the number of GitHub comments on the commit. This
 	// is only populated for requests that fetch GitHub data like
 	// Pulls.ListCommits, Repositories.ListCommits, etc.
-	CommentCount *int `json:"comment_count,omitempty"`
+	CommentCount *int    `json:"comment_count,omitempty"`
+	NodeID       *string `json:"node_id,omitempty"`
```
Comment (shurcooL) [inline github/git_commits.go]: I think it'd be nicer to either move `NodeID` field up, together with all other fields, or add a blank line after `CommentCount`, since it has a comment saying that "This is only populated for requests that fetch GitHub data", which doesn't apply to `NodeID` as I understand.

Minor, but worth doing.

File: github/git_blobs_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (shurcooL) [reply github/git_blobs_test.go]: Are you still planning to do this? Just checking.

File: github/issues.go [hunk 8]
```diff
@@ -54,6 +55,7 @@ type Issue struct {
 	// TextMatches is only populated from search results that request text matches
 	// See: search.go and https://developer.github.com/v3/search/#text-match-metadata
 	TextMatches []TextMatch `json:"text_matches,omitempty"`
+	NodeID      *string     `json:"node_id,omitempty"`
```
Comment (shurcooL) [inline github/issues.go]: Also here, move `NodeID` field up to be with all other fields, rather than with `TextMatches`.

**Comment by anubhakushwaha** [comment]
@shurcooL @sahildua2305 Apologies for the delay :smile: Updated the PR. Please review.

File: github/github-accessors.go [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2017 The go-github AUTHORS. All rights reserved.
+// Copyright 2018 The go-github AUTHORS. All rights reserved.
```
Comment (shurcooL) [inline github/github-accessors.go]: You'll need to revert this change. See #823 for the reason why. Sorry about the trouble.

Comment (shurcooL) [reply github/github-accessors.go]: I've applied this trivial change.

**Comment by shurcooL** [comment]
Merged. Thanks a lot for working on this @anubhakushwaha!

File: github/github-accessors.go (hunk 9 above)
Comment (anubhakushwaha) [reply github/github-accessors.go]: @shurcooL It was my pleasure working on this :smile:

**Comment by anubhakushwaha** [comment]
@shurcooL I would love to finish the whole issue and will soon send the PR for the rest of the API's as well
